KIPs/kip-81.md

```diff
@@ -0,0 +1,773 @@
+        uint256 voteCap = 1;
+        if (eligibleNodes > 1) {
+            voteCap = eligibleNodes - 1;
+        }
+        uint256 votes = balance / MIN_STAKE();
+        if (votes > voteCap) {
+            votes = voteCap;
+        }
+        return votes;
+    }
+
+    function refreshStake(address staking) external override {
+        for (uint256 i = 0; i < allTrackerIds.length; i++) {
+            uint256 trackerId = liveTrackerIds[i];
+            Tracker storage tracker = trackers[trackerId];
+            if (tracker.trackStart <= block.number && block.number < tracker.trackEnd) {
+                updateTracker(tracker, staking);
+            }
+        }
+    }
+
+    function updateTracker(Tracker storage tracker, address staking) private {
+        // Check that `staking` is valid GC contract
+        address nodeId = tracker.stakingToNodeId[staking];
+        if (nodeId == address(0)) {
+            return;
+        }
+
+        // Update balance
+        uint256 oldBalance = tracker.stakingBalances[staking];
+        uint256 newBalance = msg.sender.balance;
+        tracker.stakingBalances[staking] = newBalance;
+        tracker.nodeBalances[nodeId] -= oldBalance;
+        tracker.nodeBalances[nodeId] += newBalance;
+        uint256 nodeBalance = tracker.nodeBalances[nodeId];
+
+        // Update votes
+        uint256 oldVotes = tracker.nodeVotes[nodeId];
+        uint256 newVotes = calcVotes(tracker.eligibleNodes, nodeBalance);
+        tracker.nodeVotes[nodeId] = newVotes;
+        tracker.totalVotes -= oldVotes;
+        tracker.totalVotes += newVotes;
+    }
+
+    function refreshVoter(address staking) external {
+        address nodeId = resolveStakingFromAddressBook(staking);
+        require(nodeId != address(0), "not a staking contract");
+
+        // Unlink existing two-way mapping
+        address oldVoter = nodeIdToVoter[nodeId];
+        if (oldVoter != address(0)) {
+            voterToNodeId[oldVoter] = address(0);
+            nodeIdToVoter[nodeId] = address(0);
+        }
+        // Create new mapping
+        address newVoter = ICnStakingV2(staking).getVoterAddress();
+        if (newVoter != address(0)) {
+            require(voterToNodeId[newVoter] == address(0),
+                    "Voter address already taken");
+            voterToNodeId[newVoter] = nodeId;
+            nodeIdToVoter[nodeId] = newVoter;
+        }
+    }
+
+    function resolveStakingFromAddressBook(address staking)
+        private view returns(address)
+    {
+        (address[] memory nodeIds,
+         address[] memory stakingContracts,
+         address[] memory rewardAddrs, , ) = IAddressBook(ADDRESS_BOOK_ADDRESS()).getAllAddressInfo();
+
+        address rewardAddr;
+        for (uint256 i = 0; i < nodeIds.length; i++) {
+            if (stakingContracts[i] == staking) {
+                rewardAddr = rewardAddrs[i];
+                break;
+            }
+        }
+        if (rewardAddr != address(0)) {
```

Comment:
hmm it seems weird... if reward address is not zero, return zero..?!

Reply:
The condition is unnecessary. Deleted in acfd52e.